text/0000-packages-as-optional-namespaces.md

```diff
@@ -0,0 +1,198 @@
+
+The crate can be imported in Cargo.toml using its name as normal:
+
+```toml
+[dependencies]
+"foo/bar" = "1.0"
+```
+
+
+In Rust code, the slash gets converted to an underscore, the same way we do this for dashes.
+
+```rs
+use foo_bar::Baz;
+```
+
+# Reference-level explanation
+
+`/` is now considered a valid identifier inside a crate name Crates.io. For now, we will restrict crate names to having a single `/` in them, not at the beginning or end of the name, but this can be changed in the future.
+
+When publishing a crate `foo/bar`, if the crate does not exist, the following must be true:
+
+ - `foo` must exist
+ - The user publishing the crate must be an owner of `foo`
+
+For the crate `foo/bar`, all owners of `foo` are always considered owners of `foo/bar`, however additional owners may be added. People removed from ownership of `foo` will also lose access to `foo/bar` unless they were explicitly added as owners to `foo/bar`.
+
+Crates.io displays `foo/bar` crates with the name `foo/bar`, though it may stylistically make the `foo` part link to the `foo` crate.
+
+The [registry index trie](https://doc.rust-lang.org/nightly/cargo/reference/registries.html#index-format) may represent subpackages by placing `foo/bar` in `foo@/bar`, placed next to where `foo` is in the trie (i.e. the full path will be `fo/foo@/bar`).
+
+No changes are made to `rustc`. When compiling a crate `foo/bar`, Cargo will automatically pass in `--crate-name foo_bar`, and when referring to it as a dependency Cargo will use `--extern foo_bar=....`. This is the same thing we currently do for `foo-bar`.
+
+If you end up in a situation where you have both `foo/bar` and `foo-bar` as active dependencies of your crate, your code will not compile and you must [rename](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#renaming-dependencies-in-cargotoml) one of them.
+
+The `features = ` key in Cargo.toml continues parsing `foo/bar` as "the feature `bar` on dependency `foo`", however it now will unambiguously parse strings ending with a slash (`foo/` and `foo/bar/`) as referring to a dependency, as opposed to feature on a dependency. Cargo may potentially automatically handle the ambiguity or error about it.
+
+# Drawbacks
+
+## Slashes
+So far slashes as a "separator" have not existed in Rust. There may be dissonance with having another non-identifier character allowed on crates.io but not in Rust code. Dashes are already confusing for new users. Some of this can be remediated with appropriate diagnostics on when `/` is encountered at the head of a path.
+
+
+Furthermore, slashes are ambiguous in feature specifiers (though a solution has been proposed above for this):
```

Comment:
I feel like this focuses on the syntactic ambiguity but overlooks the end-user ambiguity which is pretty bad in my opinion.  A trailing slash doesn't make this any clearer.

Reply:
I support what @epage is saying. If you recall, I had already brought this up in the previous discussion at:

- https://github.com/Manishearth/namespacing-rfc/issues/1
- https://github.com/Manishearth/namespacing-rfc/issues/4

End user experience is severely affected if we move forward with `/` as a separator.